### PR TITLE
improve emit types

### LINKED ISSUE #4813: Type issues with `emit` and `emitWithAck`
**Describe the bug**
When using `emit`, `emitWithAck`, and the server side variants, there are issues with the types of Broadcaster, Server, Namespace, and Socket  types.

**To Reproduce**

Socket.IO server version: `4.7.2`

*Server*

```js
import { Server } from "socket.io";

export interface ServerToClientEvents {
  basicEmit: (a: number, b: string) => void;
  withAck: (d: string, callback: (e: number) => void) => void;
  // multiple Ack arguments doesn't work on broadcasts - single target
  withAck2: (d: string, callback: (e: number, f: string) => void) => void;
}
export interface ClientToServerEvents {}
const io = new Server<ClientToServerEvents,ServerToClientEvents>(3000, {});

//Emits
// no callback
io.timeout(1).emit('basicEmit',3,'')
// Likely should be a type error as it doesn't have a callback at all
// Currently just times out...

// One arg in callback
io.emit('withAck', 'hello', (...args) => { });
//=> ❌ typeof args === [number], should be [number[]]
io.timeout(1).emit('withAck', 'hello', (...args) => {});
//=> ✅ typeof args === [Error, number[]]
io.to('1').to('2').emit('withAck','',(...args)=>{})
//=> ❌ typeof args === [number], should be [number[]]

// Two args in callback
io.emit('withAck2', 'hello', (...args) => { });
//=> ❌ typeof args === [number,string], should be [number[]]
io.timeout(1).emit('withAck2','hello',(...args)=>{});
//=> ❌ typeof args === [Error,number,string], should be [Error,number[]]

// EmitWithAcks
// no callback
io.emitWithAck('basicEmit',3,'')
//=> type error: Expected 2 arguments, got 3 (the last argument is clipped off)
// Should just be a type-error, as without `timeout` applied, it will likely just never resolve
// if you apply a timeout, then it just times out eventually due to not having a callback

// One arg in callback
io.emitWithAck('withAck','hello');
//=>❌ type is Promise<any>, should be Promise<number[]>
io.timeout(1).emitWithAck('withAck','hello')
//=>✅ type is Promise<number[]>

// Two args in callback - doesn't even work in timeout broadcaster
io.emitWithAck('withAck2','hello');
//=>❌ type is Promise<any>, should be Promise<number[]>
io.timeout(1).emitWithAck('withAck2','hello')
//=>❌ type is Promise<any>, should be Promise<number[]>


io.on("connection", (socket) => {
  // emit is fine on the socket instance itself
  socket.broadcast.emit('withAck','hello',(...args)=>{});
  //=>❌ typeof args === [number], should be [number[]]
  socket.broadcast.timeout(1000).emit('withAck', 'hello',(...args)=>{});
  //=>✅ typeof args === [Error, number[]]
  // important to note that whatever types exist needs to work with socket.broadcast.timeout(1) AND socket.timeout(1).broadcast
  socket.timeout(1).broadcast.timeout(1).emit('withAck','hello',(...args)=>{});
  //=>❌ typeof args === [Error, Error, number[]], should be [Error, number[]]
  // Needs to not re-decorate the event if an error was already added. (not sure what benefit allowing this gives in the first place
  
  // emitWithAck is backwards here vs the broadcast based ones...
  socket.emitWithAck('withAck','hello');
  //=>✅ Promise<number>
  socket.timeout(1).emitWithAck('withAck','hello');
  //=>❌ Promise<any>, should be Promise<number>
  
  // two args
  socket.emitWithAck('withAck2','hello');
  //=>❌ Promise<any>, should be Promise<number>
  socket.timeout(1).emitWithAck('withAck2','hello');
  //=>❌ Promise<any>, should be Promise<number>
});
```

Socket.IO client version: `x.y.z`

*Client*
I haven't checked on the client, but I assume the same types are used on here too (to some extent)

**Expected behavior**
Type definitions should be consistent with what socket.io is doing so that we can fully rely on them. As of the moment, they are wrong in a few places that are either frustrating (`any` instead of a real type) or problematic (`number` instead of `number[]`).

**Platform:**
 - Typescript 5.2

**Additional context**
I have some first passes at solving some of these issues that I can work towards improving and turning into a PR, if that's desired.

### PR DESCRIPTION
**Pull request by ZachHaber**
### The kind of change this PR does introduce

* [x] a bug fix
* [ ] a new feature
* [ ] an update to the documentation
* [ ] a code change that improves performance
* [ ] other

### Current behavior
The types of the `emit`, `emitWithAck`, `serverSideEmit`, and `serverSideEmitWithAck` have issues with their types.

They revert to `any` very often or aren't arrays when they should be (server/namespace emit and server/namespace => broadcastOperator emit).

### New behavior
This fixes most of this behavior and adds tests for them.

### Other information (e.g. related issues)
Fixes #4813

### CODE REVIEW

**Comment by ZachHaber** [comment]
Open question:
- Should the `emitWithAck`/`serverSideEmitWithAck` functions only allow events names that have ack functions?

**Comment by ZachHaber** [comment]
@darrachequesne, I've improved the types further, and also removed `emitWithAck` from the server and namespace.
Some of these type changes will need to be mirrored to the client library to fix up emitWithAck types there. I also can't change the docs website to adjust it to the changes.

I'll be away from computers from Saturday till October, so feel free to make any needed changes to the PR.

**Comment by ZachHaber** [comment]
@darrachequesne, could you review this when you get a chance?

**Comment by darrachequesne** [comment]
@ZachHaber sorry for the delay, I'm taking a look at this.

File: lib/typed-events.ts [hunk 1]
```diff
@@ -241,3 +301,11 @@ export type DecorateAcknowledgementsWithMultipleResponses<E> = {
     ? (...args: ExpectMultipleResponses<Params>) => Result
     : E[K];
 };
+
+export type RemoveAcknowledgements<E> = {
```
Comment (darrachequesne) [inline lib/typed-events.ts:305]: I'm not sure to understand here, shouldn't the events with an acknowledgement be flltered out?

File: package.json [hunk 2]
```diff
@@ -52,7 +52,8 @@
     "debug": "~4.3.2",
     "engine.io": "~6.5.2",
     "socket.io-adapter": "~2.5.2",
-    "socket.io-parser": "~4.2.4"
+    "socket.io-parser": "~4.2.4",
+    "type-fest": "^3.13.1"
```
Comment (darrachequesne) [inline package.json:56]: We try to keep our prod dependencies to the bare minimum, so I think we'll indeed need to copy the types from `type-fest`.

File: lib/namespace.ts [hunk 3]
```diff
@@ -252,7 +255,10 @@ export class Namespace<
    * @return a new {@link BroadcastOperator} instance for chaining
    */
   public to(room: Room | Room[]) {
-    return new BroadcastOperator<EmitEvents, SocketData>(this.adapter).to(room);
+    return new BroadcastOperator<
+      DecorateAcknowledgementsWithMultipleResponses<EmitEvents>,
```
Comment (darrachequesne) [inline lib/namespace.ts:259]: Couldn't we move the `DecorateAcknowledgementsWithMultipleResponses` type to the `BroadcastOperator` class, so we don't have to apply it everytime?

Comment (ZachHaber) [reply lib/namespace.ts:259]: I definitely think that would be nicer; however after looking at the code, I don't think we can.
The `BroadcastOperator` has an `expectSingleResponse` flag, which changes the behavior between single responses and multiple. With that, the options I see are to either: apply the decorations when switching to a `BroadcastOperator`, or to setup the expectSingleResponse as a third generic type parameter on the BroadcastOperator.

**Comment by darrachequesne** [comment]
Awesome work, thanks a lot :+1:

**Comment by lts20050703** [comment]
@ZachHaber Awesome work, thank you for your contribution! However, it seems like your PR caused #4914, is there any chance you can look into this?

**Comment by ZachHaber** [comment]
@lts20050703 I'll take a look and see if I can figure that out!

**Pull request merged**